File: examples/cpp/server_example.cpp
#include <chiral/server.hpp>
#include <Eigen/Dense>
#include <atomic>
#include <chrono>
#include <cstdio>
#include <thread>

using Clock = std::chrono::steady_clock;
using Ms    = std::chrono::duration<double, std::milli>;

static constexpr int H   = 480;
static constexpr int W   = 640;
static constexpr int DOF = 7;

static std::vector<chiral::CameraConfig> make_cam_configs() {
    Eigen::Matrix3d K;
    K << 600,   0, 320,
           0, 600, 240,
           0,   0,   1;

    std::vector<chiral::CameraConfig> cfgs;
    for (const char* name : {"cam_0", "cam_1", "cam_2", "cam_3"}) {
        chiral::CameraConfig c;
        c.name      = name;
        c.height    = H;
        c.width     = W;
        c.channels  = 3;
        c.has_depth = true;
        c.intrinsics = K;
        c.extrinsics = Eigen::Matrix4d::Identity();
        cfgs.push_back(c);
    }
    return cfgs;
}

static std::vector<chiral::ProprioConfig> make_proprio_configs() {
    return {{"joint_pos", DOF}, {"joint_vel", DOF}};
}

class MyRobotServer : public chiral::PolicyServer {
    int               step_    = 0;
    double            t_sum_   = 0.0;
    std::atomic<bool> running_{true};

public:
    MyRobotServer(const std::string& host, int port)
        : PolicyServer(make_cam_configs(), make_proprio_configs(), host, port)
    {
        // One capture thread per camera.
        for (std::size_t i = 0; i < configs_.size(); ++i)
            std::thread(&MyRobotServer::camera_loop, this, i).detach();

        // One update thread per proprio stream.
        for (std::size_t i = 0; i < proprio_configs_.size(); ++i)
            std::thread(&MyRobotServer::proprio_loop, this, i).detach();
    }

    ~MyRobotServer() { running_ = false; }

    chiral::InfoMap get_metadata() override {
        return {{"cameras", "cam_0,cam_1,cam_2,cam_3"}, {"action_N", "1"}, {"action_D", "7"}};
    }

    std::pair<chiral::Observation, chiral::InfoMap> reset() override {
        step_ = 0; t_sum_ = 0.0;
        return {make_obs(), {}};
    }

    chiral::StepResult step(const chiral::Action& /*action*/) override {
        auto t0 = Clock::now();

        // make_obs() snapshots all camera and proprio buffers under their locks.
        chiral::StepResult r;
        r.obs       = make_obs(step_ * 0.05);
        r.reward    = 0.f;
        r.truncated = false;

        double step_ms = Ms(Clock::now() - t0).count();
        t_sum_ += step_ms;
        r.terminated = ++step_ >= 200;

        if (step_ % 10 == 0)
            std::printf("step=%4d  server_step=%5.2fms  avg=%5.2fms\n",
                        step_, step_ms, t_sum_ / step_);
        return r;
    }

private:
    void camera_loop(std::size_t idx) {
        // Simulates a sensor driver running at ~30 Hz.
        double t = 0.0;
        const auto& c = configs_[idx];
        std::vector<uint8_t> image(c.height * c.width * c.channels, 0);
        std::vector<float>   depth(c.height * c.width, 0.f);
        while (running_) {
            // Simulated RGB frame — replace with real hardware capture.
            std::fill(image.begin(), image.end(),
                      static_cast<uint8_t>(128 + 127 * std::sin(t)));
            update_image(idx, image.data(), image.size());

            // Simulated depth map (metres) — replace with real hardware capture.
            std::fill(depth.begin(), depth.end(),
                      1.0f + 0.5f * static_cast<float>(std::sin(t)));
            update_depth(idx, reinterpret_cast<const uint8_t*>(depth.data()),
                         depth.size() * sizeof(float));

            // Update extrinsics every frame for moving cameras (e.g. wrist).
            // In real code: T = fk_solver.compute(joint_positions)
            Eigen::Matrix4d T = Eigen::Matrix4d::Identity();
            T(0, 3) = 0.1 * std::sin(t);  // simulated oscillating translation
            update_extrinsics(idx, T);

            t += 1.0 / 30.0;
            std::this_thread::sleep_for(std::chrono::milliseconds(33));
        }
    }

    void proprio_loop(std::size_t idx) {
        // Simulates a proprioception driver running at ~500 Hz.
        const int n = proprio_configs_[idx].size;
        Eigen::VectorXf state = Eigen::VectorXf::Zero(n);
        while (running_) {
            // ── hardware read goes here ────────────────────────────────────
            // state = robot.read_joints();  (Eigen::VectorXf, length n)
            update_proprio(idx, state.data(), state.size());
            // ──────────────────────────────────────────────────────────────
            std::this_thread::sleep_for(std::chrono::milliseconds(2));
        }
    }
};

int main() {
    MyRobotServer("0.0.0.0", 8765).run();
}

File: examples/python/server_example.py
"""Example robot server using the streaming (decoupled) API.

Observations are served on demand via get_obs(); actions arrive
independently via apply_action() dispatched at a fixed Hz by the client.

Run first, then start client_example.py:

    uv run examples/python/server_example.py &
    uv run examples/python/client_example.py
"""
import threading
import time

import numpy as np
import chiral

H, W     = 480, 640
CAMERAS  = ["cam_0", "cam_1", "cam_2", "cam_3", "cam_4", "cam_5", "cam_6", "cam_7"]
DOF      = 7  # joint-space dimensionality

# Intrinsics / extrinsics per camera (dummy values).
INTRINSICS = np.array([[600, 0, 320], [0, 600, 240], [0, 0, 1]], dtype=np.float64)
EXTRINSICS = [np.eye(4, dtype=np.float64)] * len(CAMERAS)


class MyRobotServer(chiral.PolicyServer):
    def __init__(self):
        super().__init__(host="0.0.0.0", port=8765)
        # self.images, self.depths, self.proprios, and all locks are
        # pre-allocated by the base class.
        self._step    = 0
        self._t_sum   = 0.0
        self._running = True

        # One capture thread per camera.
        for name in self.images:
            threading.Thread(
                target=self._camera_loop, args=(name,), daemon=True
            ).start()

        # One update thread per proprio stream.
        for name in self.proprios:
            threading.Thread(
                target=self._proprio_loop, args=(name,), daemon=True
            ).start()

    def camera_configs(self) -> list[chiral.CameraConfig]:
        return [
            chiral.CameraConfig(
                name=name,
                height=H,
                width=W,
                channels=3,
                has_depth=True,
                intrinsics=INTRINSICS,
                extrinsics=EXTRINSICS[i],
            )
            for i, name in enumerate(CAMERAS)
        ]

    def proprio_configs(self) -> list[chiral.ProprioConfig]:
        return [
            chiral.ProprioConfig(name="joint_pos", size=DOF),
            chiral.ProprioConfig(name="joint_vel", size=DOF),
        ]

    def _camera_loop(self, name: str) -> None:
        """Simulates a sensor driver running at ~30 Hz per camera."""
        t = 0.0
        while self._running:
            # Simulated RGB frame and depth map — replace with real hardware capture.
            new_image = np.random.randint(0, 256, (H, W, 3), dtype=np.uint8)
            new_depth = np.random.uniform(0.5, 5.0, (H, W)).astype(np.float32)  # metres
            self.update_image(name, new_image)
            self.update_depth(name, new_depth)

            # Update extrinsics every frame for moving cameras (e.g. wrist).
            T = np.eye(4, dtype=np.float64)
            T[0, 3] = 0.1 * np.sin(t)  # simulated oscillating translation
            self.update_extrinsics(name, T)

            t += 1 / 30
            time.sleep(1 / 30)

    def _proprio_loop(self, name: str) -> None:
        """Simulates a proprioception driver running at ~500 Hz."""
        while self._running:
            # state = robot.read_joints()
            # self.update_proprio(name, state)
            time.sleep(1 / 500)

    async def get_metadata(self) -> dict:
        return {"cameras": CAMERAS, "action_shape": [1, DOF]}

    async def reset(self) -> tuple[chiral.Observation, dict]:
        self._step  = 0
        self._t_sum = 0.0
        return self._make_obs(), {}

    async def get_obs(self) -> chiral.Observation:
        """Return a snapshot of the current sensor state."""
        return self._make_obs()

    async def apply_action(self, action: np.ndarray) -> None:
        """Receive one action slice from the client and apply it to the robot."""
        t0 = time.perf_counter()

        # ── hardware command goes here ─────────────────────────────────────
        # robot.send_joint_command(action)
        # ──────────────────────────────────────────────────────────────────

        step_ms = (time.perf_counter() - t0) * 1e3
        self._step  += 1
        self._t_sum += step_ms

        print(f"action={self._step:4d}  "
              f"apply={step_ms:5.2f}ms  "
              f"avg={self._t_sum / self._step:5.2f}ms",
              flush=True)


if __name__ == "__main__":
    MyRobotServer().run()
